spec/bolt_server/app_integration_spec.rb

```diff
@@ -171,6 +176,191 @@ def app
       end
     end
 
+    describe 'apply_prep' do
+      let(:path) { "/ssh/apply_prep" }
+
+      it 'apply_prep runs install task configured in plugin_hooks and gathers custom facts' do
+        # Target a spec container that already has an agent on it
+        inventory = Bolt::Inventory.empty
+        puppet_6_agent_container = inventory.get_target(conn_uri('ssh', include_password: true, override_port: 20024))
+        target = {
+          hostname: puppet_6_agent_container.host,
+          user: puppet_6_agent_container.user,
+          password: puppet_6_agent_container.password,
+          port: puppet_6_agent_container.port,
+          plugin_hooks: { 'puppet_library' => { 'plugin' => 'task', 'task' => 'fake_puppet_agent::install',
+                                                'parameters' => {} } }
+        }
+        body = {
+          versioned_project: 'bolt_server_test_project',
+          target: target,
+          job_id: 1
+        }
+        post(path, JSON.generate(body), 'CONTENT_TYPE' => 'text/json')
+        expect(last_response).to be_ok
+        expect(last_response.status).to eq(200)
+        result = JSON.parse(last_response.body)
+        expect(result['status']).to eq('success')
+        expect(result['value'].keys).to include('os')
+      end
+
+      it 'apply_prep fails when install task fails' do
+        # Target a spec container that already has an agent on it
+        inventory = Bolt::Inventory.empty
+        puppet_6_agent_container = inventory.get_target(conn_uri('ssh', include_password: true, override_port: 20024))
+        target = {
+          hostname: puppet_6_agent_container.host,
+          user: puppet_6_agent_container.user,
+          password: puppet_6_agent_container.password,
+          port: puppet_6_agent_container.port,
+          plugin_hooks: { 'puppet_library' => { 'plugin' => 'task', 'task' => 'fake_puppet_agent::install',
+                                                'parameters' => { 'fail' => true } } }
+        }
+        body = {
+          versioned_project: 'bolt_server_test_project',
+          target: target,
+          job_id: 1
+        }
+        post(path, JSON.generate(body), 'CONTENT_TYPE' => 'text/json')
+        expect(last_response).to be_ok
+        expect(last_response.status).to eq(200)
+        result = JSON.parse(last_response.body)
+        expect(result['status']).to eq('failure')
+        expect(result['object']).to eq('fake_puppet_agent::install')
+      end
+
+      it 'apply_prep fails when target does not define suitable puppet_library plugin_hook' do
```

Comment:
Wonder if we can just expect callers to handle this or encode this in the schema somewhere...

Reply:
I thought of that too. I think it would require a new "target" schema though special for apply/apply_prep. I suppose i could do that but i'de rather just do an explicit check for now to avoid duplicating target.